### PR TITLE
Feature/poseidon query optimization

### PR DESCRIPTION
**Pull request by Kolezhniuk**
https://idyllicvision.atlassian.net/browse/IDEN-519

### CODE REVIEW

File: contracts/lib/PoseidonExtended.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                    values[
+                        iterationIndex + 5 >= values.length
+                            ? 0
+                            : iterationIndex + 5
+                    ]
+                ]
+            );
+
+
+            fullHash = poseidon2.poseidon([fullHash, poseidonHash]);
```
Comment (AndriianChestnykh) [inline contracts/lib/PoseidonExtended.sol:69]: If` i=0` then there is no reason to hash 1 with the `poseidonHash`

File: contracts/lib/PoseidonExtended.sol [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    function poseidonFold(uint256[] memory values)
+        public
+        view
+        returns (uint256)
+    {
+        uint256 moduloRest = values.length % BATCH_SIZE;
+        uint256 difftoRound = BATCH_SIZE - moduloRest;
+        uint256 fullLength = values.length + difftoRound;
+        uint256 totalIterations = fullLength / BATCH_SIZE;
+        uint256 fullHash = 1;
```
Comment (AndriianChestnykh) [inline contracts/lib/PoseidonExtended.sol:32]: I doubt we need it.

File: contracts/validators/CredentialAtomicQuerySigValidatorV2.sol [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        // verify user states
+
+        uint256 userId = inputs[USER_ID_INDEX];
+        uint256 userState = inputs[3];
+        uint256 issuerAuthState = inputs[0];
+        uint256 issuerId = inputs[5];
+        uint256 issuerClaimNonRevState = inputs[6];
+
+        // 1. User state must be latest or genesis
```
Comment (AndriianChestnykh) [inline contracts/validators/CredentialAtomicQuerySigValidatorV2.sol:91]: May be better move `User state must be latest or genesis` check to utils? We're repeating it a lot

File: contracts/validators/CredentialAtomicQuerySigValidatorV2.sol [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            );
+        } else {
+            // The non-empty state is returned, and it’s not equal to the state that the user has provided.
+            require(
+                userStateFromContract == userState,
+                "user state is not latest"
+            );
+        }
+
+        // 2. Issuer state must be registered in state contracts or be genesis
```
Comment (AndriianChestnykh) [inline contracts/validators/CredentialAtomicQuerySigValidatorV2.sol:108]: The same, may be move this whole check to utils?

File: contracts/validators/CredentialAtomicQuerySigValidatorV2.sol (hunk 3 above)
Comment (Kolezhniuk) [reply contracts/validators/CredentialAtomicQuerySigValidatorV2.sol:91]: Is this check will always be the same between Mtp and Sig contracts?

**Pull request closed without merging**